src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/Shell/HierarchyIdTests.cs

```diff
@@ -0,0 +1,177 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Xunit;
+
+namespace Microsoft.VisualStudio.Shell
+{
+    [ProjectSystemTrait]
+    public class HierarchyIdTests
+    {
+        [Fact]
+        public void Constructor1_ReturnsEmpty()
+        {
+            var hierarchyId = new HierarchyId();
+
+            Assert.True(hierarchyId.IsEmpty);
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData((uint)VSConstants.VSITEMID.Nil)]
+        [InlineData((uint)VSConstants.VSITEMID.Root)]
+        [InlineData((uint)VSConstants.VSITEMID.Selection)]
+        public void Constructor_ValueAsId_SetsId(uint id)
+        {
+            var hierarchyId = new HierarchyId(id);
+
+            Assert.Equal(id, hierarchyId.Id);
+        }
+
+        [Fact]
+        public void IsRoot_WhenIdIsRoot_ReturnsTrue()
+        {
+            var hierarchyId = new HierarchyId((uint)VSConstants.VSITEMID.Root);
+
+            Assert.True(hierarchyId.IsRoot);
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData((uint)VSConstants.VSITEMID.Nil)]
+        [InlineData((uint)VSConstants.VSITEMID.Selection)]
+        public void IsRoot_WhenIdIsNotRoot_ReturnsFalse(uint id)
+        {
+            var hierarchyId = new HierarchyId(id);
+
+            Assert.False(hierarchyId.IsRoot);
+        }
+
+        [Fact]
+        public void IsSelection_WhenIdIsSelection_ReturnsTrue()
+        {
+            var hierarchyId = new HierarchyId((uint)VSConstants.VSITEMID.Selection);
+
+            Assert.True(hierarchyId.IsSelection);
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData((uint)VSConstants.VSITEMID.Nil)]
+        [InlineData((uint)VSConstants.VSITEMID.Root)]
+        public void IsSelection_WhenIdIsNotSelection_ReturnsFalse(uint id)
+        {
+            var hierarchyId = new HierarchyId(id);
+
+            Assert.False(hierarchyId.IsSelection);
+        }
+
+        [Fact]
+        public void IsEmpty_WhenIdIsEmpty_ReturnsTrue()
+        {
+            var hierarchyId = new HierarchyId(0);
+
+            Assert.True(hierarchyId.IsEmpty);
+        }
+
+        [Theory]
+        [InlineData((uint)VSConstants.VSITEMID.Nil)]
+        [InlineData((uint)VSConstants.VSITEMID.Root)]
+        [InlineData((uint)VSConstants.VSITEMID.Selection)]
+        public void IsEmpty_WhenIdIsNotEmpty_ReturnsFalse(uint id)
+        {
+            var hierarchyId = new HierarchyId(id);
+
+            Assert.False(hierarchyId.IsEmpty);
```

Comment:
Nil and Empty means two different things. :(

Empty is the lack of a value, Nil is it has a value, but it's null.